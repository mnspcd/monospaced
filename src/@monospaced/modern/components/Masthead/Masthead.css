.mdrn .Masthead {
  align-items: center;
  display: flex;
  height: 4.5em;
  justify-content: space-between;
  margin: 0 auto;
  max-width: 56.75em;
  padding-left: 1em;
  padding-right: 0.25em;
  width: 100%;
}

@media (--larger-viewport) {
  .mdrn .Masthead {
    padding-right: 1em;
  }
}

.mdrn .Masthead-logo {
  color: var(--color-blue-9);
  display: block;
}

.mdrn .Masthead-logo:hover,
.mdrn .Masthead-logo:active {
  color: var(--color-blue-9);
}

.mdrn .Masthead--onPoster .Masthead-logo,
.mdrn .Masthead--onPoster .Masthead-logo:hover,
.mdrn .Masthead--onPoster .Masthead-logo:active {
  background: none;
  color: var(--color-white);
}

.mdrn .Masthead-menu {
  display: flex;
}

@media (--larger-viewport) {
  .mdrn .Masthead-menuButton {
    display: none;
  }
}

.mdrn .Masthead-nav {
  display: none;
}

@media (--larger-viewport) {
  .mdrn .Masthead-nav {
    display: block;
  }
}

.mdrn .Masthead-navList {
  font-family: var(--font-monospaced);
  font-variant: slashed-zero;
  list-style: none;
  margin-bottom: 0;
  margin-right: -0.5em;
  margin-top: 0;
  padding-left: 0;
}

.mdrn .Masthead-navItem {
  display: inline-block;
  margin-left: 0.75em;
}

.mdrn .Masthead-navLink {
  border-radius: 0.25em;
  padding: 0.25em 0.5em;
}

.mdrn .Masthead-navLink:hover {
  background-color: rgba(var(--color-blue-7-rgb), var(--opacity-hover));
}

.mdrn .Masthead-navLink:active {
  background-color: rgba(var(--color-blue-8-rgb), var(--opacity-active));
}

.mdrn .Masthead-navLinkActive {
  background-color: rgba(var(--color-blue-8-rgb), var(--opacity-static));
  color: var(--color-blue-8);
}
