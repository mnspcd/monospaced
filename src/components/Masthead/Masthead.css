.Masthead {
  align-items: center;
  display: flex;
  height: 3.5em;
  justify-content: space-between;
  margin: 0 auto;
  max-width: 100%;
  padding-left: 1em;
  padding-right: 0.25em;
  width: 56.75em;
}

@media (--larger-viewport) {
  .Masthead {
    padding-right: 1em;
  }
}

.Masthead-logo {
  color: var(--mnspcd-color-blue-9);
  display: block;
}

.Masthead-logo:hover,
.Masthead-logo:active {
  color: var(--mnspcd-color-blue-9);
}

.Masthead--onPoster .Masthead-logo,
.Masthead--onPoster .Masthead-logo:hover,
.Masthead--onPoster .Masthead-logo:active {
  background: none;
  color: var(--mnspcd-color-white);
}

.Masthead-menu {
  display: flex;
}

@media (--larger-viewport) {
  .Masthead-menuButton {
    display: none;
  }
}

.Masthead-nav {
  display: none;
}

@media (--larger-viewport) {
  .Masthead-nav {
    display: block;
  }
}

.Masthead-navList {
  font-family: var(--font-monospaced);
  font-variant: slashed-zero;
  list-style: none;
  margin-bottom: 0;
  margin-right: -0.5em;
  margin-top: 0;
  padding-left: 0;
}

.Masthead-navItem {
  display: inline-block;
  margin-left: 0.75em;
}

.Masthead-navLink {
  border-radius: 0.25em;
  padding: 0.25em 0.5em;
}

.Masthead-navLink:hover {
  background-color: rgba(var(--mnspcd-color-blue-7-rgb), var(--opacity-hover));
}

.Masthead-navLink:active {
  background-color: rgba(var(--mnspcd-color-blue-8-rgb), var(--opacity-active));
}

.Masthead-navLinkActive {
  background-color: rgba(var(--mnspcd-color-blue-8-rgb), var(--opacity-static));
  color: var(--mnspcd-color-blue-8);
}

.Masthead-navItem:last-child .Masthead-navLink {
  border: calc(1rem / 16) solid var(--mnspcd-color-blue-2);
}

.Masthead-navItem:last-child .Masthead-navLink:hover {
  border-color: var(--mnspcd-color-blue-3;)
}

.Masthead-navItem:last-child .Masthead-navLink:active {
  border-color: var(--mnspcd-color-blue-4;)
}
