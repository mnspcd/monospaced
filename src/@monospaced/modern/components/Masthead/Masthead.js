import classNames from "classnames";
import PropTypes from "prop-types";
import React, { useEffect, useState } from "react";
import { Mail, Menu, X } from "react-feather";
import { Link } from "react-router";

import Button from "../Button";
import Heading from "../Heading";
import { PosterConsumer } from "../Poster";

import "./Masthead.css";

/**
 * Use `Masthead` to display a logo and navigation links at the top of a page.
 */
const Masthead = ({
  activePath = null,
  headingLevel = null,
  links = null,
  logo,
}) => {
  const [menuOpen, setMenuOpen] = useState(false);

  useEffect(() => {
    if (menuOpen) {
      document.body.classList.add("mdrn-is-overlayed");
    } else {
      document.body.classList.remove("mdrn-is-overlayed");
    }
  }, [menuOpen]);

  return (
    <PosterConsumer>
      {poster => (
        <div
          className={classNames({
            Masthead: true,
            [`Masthead--menuOpen`]: menuOpen,
            [`Masthead--onPoster`]: poster,
          })}
        >
          <Heading level={headingLevel}>
            <Link
              className="Masthead-logo"
              onClick={() => {
                setMenuOpen(false);
              }}
              to="/"
            >
              {logo}
            </Link>
          </Heading>
          <div className="Masthead-menu">
            <div className="Masthead-menuButton">
              {links &&
                (links.length > 1 ? (
                  <Button
                    icon={
                      menuOpen ? (
                        <X aria-label="Close" role="img" size="100%" />
                      ) : (
                        <Menu aria-label="Menu" role="img" size="100%" />
                      )
                    }
                    onClick={() => {
                      setMenuOpen(!menuOpen);
                    }}
                    styleVariant="flat"
                  />
                ) : (
                  <a href={links[0].href}>
                    <Button
                      icon={<Mail aria-label="Email" role="img" size="100%" />}
                      styleVariant="flat"
                    />
                  </a>
                ))}
            </div>
            {links && (
              <nav className="Masthead-nav">
                <ul className="Masthead-navList">
                  {links.map(link => {
                    const { href, routerLink, text } = link;

                    return (
                      <li className="Masthead-navItem" key={text}>
                        {routerLink ? (
                          <Link
                            className={classNames({
                              [`Masthead-navLink`]: true,
                              [`Masthead-navLinkActive`]: href === activePath,
                            })}
                            to={href}
                          >
                            {text}
                          </Link>
                        ) : (
                          <a className="Masthead-navLink" href={href}>
                            {text}
                          </a>
                        )}
                      </li>
                    );
                  })}
                </ul>
              </nav>
            )}
          </div>
        </div>
      )}
    </PosterConsumer>
  );
};

Masthead.propTypes = {
  /**
   * Path of the current URL.
   */
  activePath: PropTypes.string,

  /**
   * HTML section heading level for the Masthead logo. If not provided will
   * render in a span.
   */
  headingLevel: PropTypes.oneOf(Heading.levels),

  /**
   * Navigation links to display in the Masthead.
   */
  links: PropTypes.arrayOf(
    PropTypes.shape({
      href: PropTypes.string.isRequired,
      routerLink: PropTypes.bool,
      text: PropTypes.string.isRequired,
    }),
  ),

  /**
   * Logo to display in the Masthead, usually a Logotype component.
   */
  logo: PropTypes.element.isRequired,
};

export default Masthead;
