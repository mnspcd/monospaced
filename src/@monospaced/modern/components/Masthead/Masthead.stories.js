import { boolean, number, select, text } from "@storybook/addon-knobs";
import React from "react";
import { Route, Router } from "react-router";
import createMemoryHistory from "react-router/lib/createMemoryHistory";

import Heading from "../Heading";
import Logotype from "../Logotype";
import Poster from "../Poster";

import Masthead from "./Masthead";

const history = createMemoryHistory();

const RouterDecorator = story => {
  const route = <Route component={() => story()} path="*" />;

  class RenderRouter extends React.Component {
    render() {
      return <Router history={history}>{route}</Router>;
    }
  }

  return <RenderRouter />;
};

export default {
  component: Masthead,
  decorators: [RouterDecorator],
  title: "Components/Masthead",
};

export const basic = () => (
  <Masthead
    links={[
      {
        href: "https://example.com",
        text: `item-1`,
      },
      {
        href: "https://example.com",
        text: `item-2`,
      },
      {
        href: "https://example.com",
        text: `item-3`,
      },
    ]}
    logo={<Logotype height="1.5em" />}
  />
);

export const onPoster = () => (
  <Poster>
    <Masthead
      links={[
        {
          href: "https://example.com",
          text: `item-1`,
        },
        {
          href: "https://example.com",
          text: `item-2`,
        },
        {
          href: "https://example.com",
          text: `item-3`,
        },
      ]}
      logo={<Logotype height="1.5em" />}
    />
  </Poster>
);

export const knobs = () => {
  const Element = boolean("on Poster", false, "Story") ? Poster : "div";

  const numberOfItems = number(
    "number of links",
    3,
    {
      range: true,
      min: 1,
      max: 5,
    },
    "Story",
  );

  return (
    <Element>
      <Masthead
        headingLevel={
          select("headingLevel", [""].concat(Heading.levels), "", "Masthead") ||
          null
        }
        links={[...Array(numberOfItems).keys()].map(i => ({
          href: text("link href", "https://example.com", "Masthead"),
          text: `${text("link text", "item", "Masthead")}-${i + 1}`,
        }))}
        logo={<Logotype height="1.5em" />}
      />
    </Element>
  );
};

knobs.story = { parameters: { docs: { disable: true } } };
