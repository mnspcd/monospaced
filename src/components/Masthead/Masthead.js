import { Link } from "react-router";
import PropTypes from "prop-types";
import React from "react";
import classNames from "classnames";

import Button from "../Button";
import { Heading } from "../Markdown";
import { Mail, Menu } from "react-feather";

import "./Masthead.css";

const Masthead = ({ activePath, headingLevel, links, logo }) => {
  return (
    <div className="Masthead">
      <Heading level={headingLevel}>
        <Link className="Masthead-logo" to="/">
          {logo}
        </Link>
      </Heading>
      <div className="Masthead-menu">
        <div className="Masthead-menuButton">
          {links &&
            (links.length > 1 ? (
              <Button
                icon={<Menu aria-label="Menu" role="img" size="100%" />}
                styleVariant="flat"
              />
            ) : (
              <a href={links[0].href}>
                <Button
                  icon={<Mail aria-label="Email" role="img" size="100%" />}
                  styleVariant="flat"
                />
              </a>
            ))}
        </div>
        {links && (
          <nav className="Masthead-nav">
            <ul className="Masthead-navList">
              {links.map(link => {
                const { href, routerLink, text } = link;

                return (
                  <li className="Masthead-navItem" key={text}>
                    {routerLink ? (
                      <Link
                        className={classNames({
                          [`Masthead-navLink`]: true,
                          [`Masthead-navLinkActive`]: href === activePath,
                        })}
                        to={href}
                      >
                        {text}
                      </Link>
                    ) : (
                      <a className="Masthead-navLink" href={href}>
                        {text}
                      </a>
                    )}
                  </li>
                );
              })}
            </ul>
          </nav>
        )}
      </div>
    </div>
  );
};

Masthead.propTypes = {
  activePath: PropTypes.string,
  headingLevel: PropTypes.oneOf(Heading.levels),
  links: PropTypes.arrayOf(
    PropTypes.shape({
      href: PropTypes.string.isRequired,
      routerLink: PropTypes.bool,
      text: PropTypes.string.isRequired,
    }),
  ),
  logo: PropTypes.element.isRequired,
};

export default Masthead;
