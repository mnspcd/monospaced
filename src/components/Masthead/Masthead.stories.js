import React from "react";
import { number, select, text } from "@storybook/addon-knobs";
import { storiesOf } from "@storybook/react";

import Masthead from "./Masthead";

import { Heading } from "../Markdown";
import Logotype from "../Logotype";

storiesOf("Components/Masthead", module).add("Component", () => {
  const numberOfItems = number(
    "number of links",
    3,
    {
      range: true,
      min: 1,
      max: 5,
    },
    "Story",
  );

  return (
    <div>
      <Masthead
        headingLevel={
          select("headingLevel", [""].concat(Heading.levels), "", "Masthead") ||
          null
        }
        links={[...Array(numberOfItems).keys()].map(i => ({
          href: text("link href", "https://example.com", "Masthead"),
          text: `${text("link text", "item", "Masthead")}-${i + 1}`,
        }))}
        logo={<Logotype height="1.5em" />}
      />
    </div>
  );
});
