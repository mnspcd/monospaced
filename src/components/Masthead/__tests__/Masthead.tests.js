/* eslint-env jest */
import React from "react";
import { render } from "enzyme";

import Masthead from "../Masthead";

describe("Masthead component", () => {
  it("should render correctly", () => {
    const component = render(<Masthead logo={<div />} />);
    expect(component).toMatchSnapshot();
  });

  it("should render heading level correctly", () => {
    const component = render(<Masthead headingLevel="1" logo={<div />} />);
    expect(component).toMatchSnapshot();
  });

  it("should render single link correctly", () => {
    const component = render(
      <Masthead
        links={[
          {
            href: "href",
            text: "text",
          },
        ]}
        logo={<div />}
      />,
    );
    expect(component).toMatchSnapshot();
  });

  it("should render links correctly", () => {
    const component = render(
      <Masthead
        links={[
          {
            href: "href",
            routerLink: true,
            text: "text",
          },
          {
            href: "href",
            text: "text",
          },
        ]}
        logo={<div />}
      />,
    );
    expect(component).toMatchSnapshot();
  });
});
