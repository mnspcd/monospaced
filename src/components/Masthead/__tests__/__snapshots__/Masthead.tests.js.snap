// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Masthead component should render correctly 1`] = `
<div
  class="Masthead"
>
  <span
    class="Heading"
  >
    <a
      class="Masthead-logo"
      href="/"
    >
      <div />
    </a>
  </span>
  <div
    class="Masthead-menu"
  >
    <div
      class="Masthead-menuButton"
    />
  </div>
</div>
`;

exports[`Masthead component should render heading level correctly 1`] = `
<div
  class="Masthead"
>
  <h1
    class="Heading"
  >
    <a
      class="Masthead-logo"
      href="/"
    >
      <div />
    </a>
  </h1>
  <div
    class="Masthead-menu"
  >
    <div
      class="Masthead-menuButton"
    />
  </div>
</div>
`;

exports[`Masthead component should render links correctly 1`] = `
<div
  class="Masthead"
>
  <span
    class="Heading"
  >
    <a
      class="Masthead-logo"
      href="/"
    >
      <div />
    </a>
  </span>
  <div
    class="Masthead-menu"
  >
    <div
      class="Masthead-menuButton"
    >
      <button
        class="Button Button--styleFlat Button--withIcon Button--iconOnly"
      >
        <span
          class="Button-icon"
        >
          <svg
            aria-label="Menu"
            fill="none"
            height="100%"
            role="img"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="100%"
            xmlns="http://www.w3.org/2000/svg"
          >
            <line
              x1="3"
              x2="21"
              y1="12"
              y2="12"
            />
            <line
              x1="3"
              x2="21"
              y1="6"
              y2="6"
            />
            <line
              x1="3"
              x2="21"
              y1="18"
              y2="18"
            />
          </svg>
        </span>
      </button>
    </div>
    <nav
      class="Masthead-nav"
    >
      <ul
        class="Masthead-navList"
      >
        <li
          class="Masthead-navItem"
        >
          <a
            class="Masthead-navLink"
          >
            text
          </a>
        </li>
        <li
          class="Masthead-navItem"
        >
          <a
            class="Masthead-navLink"
            href="href"
          >
            text
          </a>
        </li>
      </ul>
    </nav>
  </div>
</div>
`;

exports[`Masthead component should render single link correctly 1`] = `
<div
  class="Masthead"
>
  <span
    class="Heading"
  >
    <a
      class="Masthead-logo"
      href="/"
    >
      <div />
    </a>
  </span>
  <div
    class="Masthead-menu"
  >
    <div
      class="Masthead-menuButton"
    >
      <form
        action="href"
      >
        <button
          class="Button Button--styleFlat Button--withIcon Button--iconOnly"
        >
          <span
            class="Button-icon"
          >
            <svg
              aria-label="Email"
              fill="none"
              height="100%"
              role="img"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="100%"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M4 4h16c1.1 0 2 .9 2 2v12c0 1.1-.9 2-2 2H4c-1.1 0-2-.9-2-2V6c0-1.1.9-2 2-2z"
              />
              <polyline
                points="22,6 12,13 2,6"
              />
            </svg>
          </span>
        </button>
      </form>
    </div>
    <nav
      class="Masthead-nav"
    >
      <ul
        class="Masthead-navList"
      >
        <li
          class="Masthead-navItem"
        >
          <a
            class="Masthead-navLink"
            href="href"
          >
            text
          </a>
        </li>
      </ul>
    </nav>
  </div>
</div>
`;
